s3dlio : FileSystemConfig Type Mismatch in API

# s3dlio Bug Report: FileSystemConfig Type Mismatch in API

## Summary
The `s3dlio` library (v0.9.7) has a type mismatch between the exported `FileSystemConfig` in `api.rs` and the one actually used by `store_for_uri_with_config()` in `object_store.rs`. This makes the public API unusable for configuring page cache behavior on file:// URIs.

**Impact**: Users cannot use the advertised `store_for_uri_with_config()` function with `FileSystemConfig` for page cache control.

**Severity**: Medium - Feature is advertised but broken; workaround exists by using `FileSystemObjectStore::with_config()` directly.

## Bug Details

### Current State (v0.9.7, commit c21d46d)

**File: `src/api.rs` (line 56)**
```rust
/// FileSystem-specific configuration (page cache, range engine, etc.)
pub use crate::file_store::FileSystemConfig;

/// Create an object store with FileSystemConfig (for page cache control, range engine, etc.)
pub use crate::object_store::store_for_uri_with_config;
```

**File: `src/object_store.rs` (lines 45, 1819-1820)**
```rust
// Line 45 - Import
use crate::file_store_direct::{ConfigurableFileSystemObjectStore, FileSystemConfig};

// Lines 1819-1820 - Function signature
pub fn store_for_uri_with_config(uri: &str, file_config: Option<FileSystemConfig>) -> Result<Box<dyn ObjectStore>> {
    store_for_uri_with_config_and_logger(uri, file_config, None)
}
```

### The Problem

There are **TWO different `FileSystemConfig` types** in s3dlio:

1. **`crate::file_store::FileSystemConfig`** (lines 32-45 of `src/file_store.rs`)
   - Purpose: Configure regular file:// URIs
   - Fields:
     - `enable_range_engine: bool`
     - `range_engine: RangeEngineConfig`
     - `page_cache_mode: Option<PageCacheMode>` ← **This is what users need for fadvise**
   - Exported from `api.rs`

2. **`crate::file_store_direct::FileSystemConfig`** (lines 32-42 of `src/file_store_direct.rs`)
   - Purpose: Configure direct:// URIs with O_DIRECT
   - Fields:
     - `direct_io: bool`
     - `alignment: usize`
     - `min_io_size: usize`
     - `sync_writes: bool`
   - **Does NOT have `page_cache_mode` field**
   - Used internally by `store_for_uri_with_config()`

### Compilation Error When Using Public API

```rust
use s3dlio::FileSystemConfig;  // Gets file_store::FileSystemConfig from api.rs
use s3dlio::PageCacheMode;
use s3dlio::object_store::store_for_uri_with_config;

let config = FileSystemConfig {
    enable_range_engine: false,
    range_engine: Default::default(),
    page_cache_mode: Some(PageCacheMode::Sequential),
};

// This fails to compile!
let store = store_for_uri_with_config("file:///data/test/", Some(config))?;
```

**Error:**
```
error[E0308]: mismatched types
expected `s3dlio::file_store_direct::FileSystemConfig`
found `s3dlio::file_store::FileSystemConfig`
```

## Root Cause Analysis

The `store_for_uri_with_config()` function implementation (lines 1819-1847 in `object_store.rs`) uses the wrong `FileSystemConfig` type:

```rust
// Line 45 - This import shadows the exported type!
use crate::file_store_direct::{ConfigurableFileSystemObjectStore, FileSystemConfig};

// Lines 1833-1846 - Uses file_store_direct::FileSystemConfig
Scheme::File => {
    if let Some(config) = file_config {
        ConfigurableFileSystemObjectStore::boxed(config)
    } else {
        FileSystemObjectStore::boxed()
    }
}
Scheme::Direct => {
    if let Some(config) = file_config {
        ConfigurableFileSystemObjectStore::boxed(config)
    } else {
        ConfigurableFileSystemObjectStore::boxed_direct_io()
    }
}
```

**Problem**: Both `Scheme::File` and `Scheme::Direct` branches use `ConfigurableFileSystemObjectStore`, which expects `file_store_direct::FileSystemConfig`. But `file://` URIs should use `file_store::FileSystemConfig` with page cache control!

## Expected Behavior

The function should:
1. Accept `file_store::FileSystemConfig` for `file://` URIs
2. Accept `file_store_direct::FileSystemConfig` for `direct://` URIs
3. OR: Have separate parameters/functions for each type
4. OR: Use a unified enum/trait-based config system

## Proposed Fix (Option 1: Separate Config Types)

```rust
// In src/object_store.rs
use crate::file_store::{FileSystemObjectStore, FileSystemConfig as FileConfig};
use crate::file_store_direct::{ConfigurableFileSystemObjectStore, FileSystemConfig as DirectConfig};

pub fn store_for_uri_with_config(
    uri: &str, 
    file_config: Option<FileConfig>,
    direct_config: Option<DirectConfig>
) -> Result<Box<dyn ObjectStore>> {
    match infer_scheme(uri) {
        Scheme::File => {
            if let Some(config) = file_config {
                Ok(Box::new(FileSystemObjectStore::with_config(config)))
            } else {
                FileSystemObjectStore::boxed()
            }
        }
        Scheme::Direct => {
            if let Some(config) = direct_config {
                ConfigurableFileSystemObjectStore::boxed(config)
            } else {
                ConfigurableFileSystemObjectStore::boxed_direct_io()
            }
        }
        // ... other schemes
    }
}
```

## Proposed Fix (Option 2: Enum-based Config)

```rust
// In src/lib.rs or src/config.rs
pub enum StorageConfig {
    File(crate::file_store::FileSystemConfig),
    Direct(crate::file_store_direct::FileSystemConfig),
    Gcs(GcsConfig),
    // ... other backend configs
}

// In src/object_store.rs
pub fn store_for_uri_with_config(
    uri: &str, 
    config: Option<StorageConfig>
) -> Result<Box<dyn ObjectStore>> {
    match (infer_scheme(uri), config) {
        (Scheme::File, Some(StorageConfig::File(cfg))) => {
            Ok(Box::new(FileSystemObjectStore::with_config(cfg)))
        }
        (Scheme::File, None) => FileSystemObjectStore::boxed(),
        (Scheme::Direct, Some(StorageConfig::Direct(cfg))) => {
            ConfigurableFileSystemObjectStore::boxed(cfg)
        }
        // ... other combinations with validation
    }
}
```

## Current Workaround (sai3-bench v0.6.8)

Instead of using the broken `store_for_uri_with_config()`, directly instantiate the store:

```rust
use s3dlio::file_store::{FileSystemObjectStore, FileSystemConfig};
use s3dlio::PageCacheMode;

let config = FileSystemConfig {
    enable_range_engine: false,
    range_engine: Default::default(),
    page_cache_mode: Some(PageCacheMode::Sequential),
};

let store = FileSystemObjectStore::with_config(config);
// Use store with ObjectStore trait methods
```

## Testing Steps to Verify Fix

1. **Compile Test**: Create a test that uses the public API as documented
   ```rust
   #[test]
   fn test_file_config_public_api() {
       use s3dlio::FileSystemConfig;
       use s3dlio::PageCacheMode;
       use s3dlio::object_store::store_for_uri_with_config;
       
       let config = FileSystemConfig {
           enable_range_engine: false,
           range_engine: Default::default(),
           page_cache_mode: Some(PageCacheMode::Sequential),
       };
       
       let store = store_for_uri_with_config("file:///tmp/test/", Some(config))
           .expect("Should create file store with config");
       
       // Verify it's the right type and works
       assert!(std::any::TypeId::of_val(&*store) != std::any::TypeId::of::<()>());
   }
   ```

2. **Runtime Test**: Verify page cache hints are actually applied
   ```bash
   # Run a workload with Sequential mode
   strace -e fadvise64 ./target/release/sai3-bench run --config test_sequential.yaml 2>&1 | grep POSIX_FADV_SEQUENTIAL
   
   # Run a workload with Random mode
   strace -e fadvise64 ./target/release/sai3-bench run --config test_random.yaml 2>&1 | grep POSIX_FADV_RANDOM
   ```

3. **Integration Test**: Create test YAML configs and verify they work end-to-end

## Affected Code Locations

- `src/api.rs`: Line 56 (exports wrong type)
- `src/object_store.rs`: Lines 45, 1819-1847 (uses wrong type)
- `src/file_store.rs`: Lines 32-45, 246-249 (correct type definition and usage)
- `src/file_store_direct.rs`: Lines 32-42 (O_DIRECT specific type)

## Backward Compatibility Considerations

### Breaking Change
Any fix will be a **breaking change** if:
- The function signature of `store_for_uri_with_config()` changes
- The exported `FileSystemConfig` type changes

### Migration Path
For **Option 1** (separate parameters):
```rust
// Old (broken) - wouldn't compile anyway
store_for_uri_with_config(uri, Some(config))

// New
store_for_uri_with_config(uri, Some(file_config), None)  // for file://
store_for_uri_with_config(uri, None, Some(direct_config))  // for direct://
```

For **Option 2** (enum):
```rust
// Old (broken)
store_for_uri_with_config(uri, Some(config))

// New
store_for_uri_with_config(uri, Some(StorageConfig::File(config)))
```

## Additional Notes

- This bug was introduced when `page_cache_mode` was added to `file_store::FileSystemConfig` in v0.9.7
- The `file_store_direct::FileSystemConfig` is older and predates the page cache feature
- The naming collision between the two types was not caught during code review
- The API documentation in `api.rs` suggests the feature works, which is misleading

## Related Files for Context

- `src/page_cache.rs`: Implementation of posix_fadvise hints
- `src/file_store.rs`: Regular file:// backend with page cache support
- `src/file_store_direct.rs`: O_DIRECT file:// and direct:// backend
- `tests/test_file_range_engine.rs`: Existing tests (don't cover this API path)

## Version Information

- **s3dlio Version**: v0.9.7
- **Git Commit**: c21d46d6231a7da95b27cf6f0724ea81a414be7a
- **Discovery Date**: October 15, 2025
- **Discovered By**: Integration work in sai3-bench v0.6.8
- **Platform**: Linux (page cache hints are Linux/Unix specific)

## Priority Justification

**Medium Priority** because:
- ✅ Feature is advertised in API but doesn't work
- ✅ Workaround exists (direct instantiation)
- ❌ Not a critical security or data corruption issue
- ❌ Page cache optimization is performance enhancement, not core functionality
- ✅ Affects API design and user experience

## Success Criteria

Fix is successful when:
1. ✅ Public API compiles without type errors
2. ✅ Page cache hints are actually applied (verifiable with strace)
3. ✅ Both file:// and direct:// URIs work correctly
4. ✅ Existing tests still pass
5. ✅ New test specifically for this API path passes
6. ✅ Documentation matches implementation

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -3,7 +3,7 @@ members = [".", "crates/s3dlio-oplog"]
 
 # Workspace-level package metadata that can be inherited by all members
 [workspace.package]
-version = "0.9.31"
+version = "0.9.32"
 edition = "2021"
 authors = ["Russ Fellows"]
 license = "Apache-2.0"

--- a/README.md
+++ b/README.md
@@ -1,9 +1,9 @@
 # s3dlio - Universal Storage I/O Library
 
 [![Build Status](https://img.shields.io/badge/build-passing-brightgreen)](https://github.com/russfellows/s3dlio)
-[![Tests](https://img.shields.io/badge/tests-175%20passing-brightgreen)](docs/Changelog.md)
-[![Rust Tests](https://img.shields.io/badge/rust%20tests-175%2F175-brightgreen)](docs/Changelog.md)
-[![Version](https://img.shields.io/badge/version-0.9.31-blue)](https://github.com/russfellows/s3dlio/releases)
+[![Tests](https://img.shields.io/badge/tests-182%20passing-brightgreen)](docs/Changelog.md)
+[![Rust Tests](https://img.shields.io/badge/rust%20tests-182%2F182-brightgreen)](docs/Changelog.md)
+[![Version](https://img.shields.io/badge/version-0.9.32-blue)](https://github.com/russfellows/s3dlio/releases)
 [![PyPI](https://img.shields.io/pypi/v/s3dlio)](https://pypi.org/project/s3dlio/)
 [![License](https://img.shields.io/badge/license-Apache--2.0-blue)](LICENSE)
 [![Rust](https://img.shields.io/badge/rust-1.91%2B-orange)](https://www.rust-lang.org)
@@ -18,6 +18,16 @@ pip install s3dlio
 ```
 
 ## 🌟 Latest Release
+
+### v0.9.32 - Bug Fix: FileSystemConfig Type Mismatch (December 2025)
+
+**🐛 Fixed Issue #85**
+- Resolved type collision between `file_store::FileSystemConfig` and `file_store_direct::FileSystemConfig`
+- Introduced unified `StorageConfig` enum for type-safe configuration
+- Added debug logging for page cache mode application
+- 7 new comprehensive tests for API configuration scenarios
+
+**⚠️ Breaking Change**: `store_for_uri_with_config()` now requires `StorageConfig` wrapper
 
 ### v0.9.30 - Zero-Copy Refactor & PyO3 0.27 (December 2025)
 

--- a/docs/Changelog.md
+++ b/docs/Changelog.md
@@ -1,5 +1,66 @@
 # s3dlio Changelog
 
+## Version 0.9.32 - Bug Fix: FileSystemConfig Type Mismatch (December 2025)
+
+### 🐛 **Fixed Issue #85: FileSystemConfig Type Mismatch**
+
+**Problem**: The public API exported `crate::file_store::FileSystemConfig` (with `page_cache_mode`) but `store_for_uri_with_config()` used `crate::file_store_direct::FileSystemConfig` (with `direct_io`, `alignment`, etc.). This made it impossible to configure page cache behavior via the public API.
+
+**Solution**: Introduced unified `StorageConfig` enum:
+
+```rust
+pub enum StorageConfig {
+    File(FileSystemConfig),      // For file:// URIs
+    Direct(DirectFileSystemConfig),  // For direct:// URIs
+}
+```
+
+**⚠️ BREAKING CHANGE**: `store_for_uri_with_config()` signature changed:
+
+```rust
+// OLD (v0.9.31 and earlier)
+store_for_uri_with_config(uri, Some(config))
+
+// NEW (v0.9.32+)
+store_for_uri_with_config(uri, Some(StorageConfig::File(config)))
+store_for_uri_with_config(uri, Some(StorageConfig::Direct(config)))
+```
+
+**Migration**:
+- Import `StorageConfig` from `s3dlio::api`
+- Wrap `FileSystemConfig` in `StorageConfig::File(...)` for file:// URIs
+- Wrap `DirectFileSystemConfig` in `StorageConfig::Direct(...)` for direct:// URIs
+- Type checking now prevents using wrong config with wrong URI scheme
+
+**Improvements**:
+- ✅ Public API now compiles and works as documented
+- ✅ Page cache hints (Sequential, Random, DontNeed) now functional
+- ✅ Added debug logging to show page cache mode being applied
+- ✅ Clear error messages if wrong config type used with URI scheme
+- ✅ Comprehensive tests for all configuration scenarios
+
+**New Exports**:
+- `pub use crate::api::StorageConfig` - Unified config enum
+- `pub use crate::file_store_direct::FileSystemConfig as DirectFileSystemConfig` - O_DIRECT config
+
+**Example**:
+
+```rust
+use s3dlio::api::{FileSystemConfig, StorageConfig, PageCacheMode, store_for_uri_with_config};
+
+let config = FileSystemConfig {
+    enable_range_engine: false,
+    range_engine: Default::default(),
+    page_cache_mode: Some(PageCacheMode::Sequential),
+};
+
+// This now works!
+let store = store_for_uri_with_config(
+    "file:///data/test/", 
+    Some(StorageConfig::File(config))
+)?;
+```
+
 ## Version 0.9.31 - Maintenance Release (December 2025)
 
 - Changed license to Apache-2.0

--- a/examples/issue_85_reproduction.rs
+++ b/examples/issue_85_reproduction.rs
@@ -1,0 +1,72 @@
+// examples/issue_85_reproduction.rs
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// SPDX-FileCopyrightText: 2025 Russ Fellows <russ.fellows@gmail.com>
+
+//! Reproduction of Issue #85: FileSystemConfig Type Mismatch
+//!
+//! This example demonstrates the bug that existed in v0.9.31 and earlier:
+//! - The public API exported `crate::file_store::FileSystemConfig`
+//! - But `store_for_uri_with_config()` expected `crate::file_store_direct::FileSystemConfig`
+//! - Result: Code using the documented public API would not compile
+//!
+//! **Before Fix (v0.9.31)**: This would fail with:
+//! ```
+//! error[E0308]: mismatched types
+//! expected `s3dlio::file_store_direct::FileSystemConfig`
+//! found `s3dlio::file_store::FileSystemConfig`
+//! ```
+//!
+//! **After Fix (v0.9.32)**: This compiles and runs successfully!
+
+use anyhow::Result;
+use std::fs::File;
+use std::io::Write;
+use tempfile::TempDir;
+
+// This is what users would naturally write based on the public API documentation
+use s3dlio::api::{FileSystemConfig, PageCacheMode, StorageConfig, store_for_uri_with_config};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    println!("Issue #85 Reproduction Test");
+    println!("============================\n");
+    
+    // Create a test file
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Hello from Issue #85 test!")?;
+    drop(file);
+    
+    println!("✓ Created test file: {}", test_file.display());
+    
+    // Create FileSystemConfig with page_cache_mode - the feature users want to use
+    let config = FileSystemConfig {
+        enable_range_engine: false,
+        range_engine: Default::default(),
+        page_cache_mode: Some(PageCacheMode::Sequential),
+    };
+    
+    println!("✓ Created FileSystemConfig with PageCacheMode::Sequential");
+    
+    // Before Fix (v0.9.31): This line would cause a compile error!
+    // After Fix (v0.9.32): This works!
+    let uri = format!("file://{}", test_file.display());
+    let store = store_for_uri_with_config(&uri, Some(StorageConfig::File(config)))?;
+    
+    println!("✓ Successfully created store with config (this would have FAILED in v0.9.31)");
+    
+    // Verify it actually works
+    let data = store.get(&uri).await?;
+    assert_eq!(&data[..], b"Hello from Issue #85 test!");
+    
+    println!("✓ Successfully read data: {:?}", String::from_utf8_lossy(&data));
+    
+    println!("\n✅ SUCCESS! Issue #85 is FIXED!");
+    println!("\nIn v0.9.31 and earlier, this example would not even compile.");
+    println!("The compiler would complain about mismatched FileSystemConfig types.");
+    println!("Now it compiles and runs perfectly!\n");
+    
+    Ok(())
+}

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,15 +1,30 @@
 [project]
 name = "s3dlio"
-version = "0.9.31"
+version = "0.9.32"
 description = "High-performance S3, Azure, GCS, and file system operations for AI/ML"
 readme = "README.md"
 requires-python = ">=3.12"
-dependencies = []
+dependencies = [
+    "numpy>=2.0.0",
+]
 license = { text = "Apache-2.0" }
 authors = [{ name = "Russ Fellows", email = "russ.fellows@gmail.com" }]
 
+[project.optional-dependencies]
+torch = ["torch>=2.0.0"]
+jax = ["jax>=0.4.0", "jaxlib>=0.4.0"]
+tensorflow = ["tensorflow>=2.16.0"]
+all = ["torch>=2.0.0", "jax>=0.4.0", "jaxlib>=0.4.0", "tensorflow>=2.16.0"]
+dev = [
+    "pytest>=8.0.0",
+    "pytest-asyncio>=0.23.0",
+    "h5py>=3.0.0",
+    "maturin>=1.0.0",
+    "patchelf>=0.17.0",  # Required for maturin on Linux
+]
+
 [build-system]
-requires = ["maturin>=0.12"]
+requires = ["maturin>=1.0.0"]
 build-backend = "maturin"
 
 [tool.maturin]

--- a/src/api.rs
+++ b/src/api.rs
@@ -59,8 +59,24 @@ pub use crate::object_store::ObjectMetadata;
 /// FileSystem-specific configuration (page cache, range engine, etc.)
 pub use crate::file_store::FileSystemConfig;
 
+/// Direct I/O FileSystem configuration (O_DIRECT, alignment, etc.)
+pub use crate::file_store_direct::FileSystemConfig as DirectFileSystemConfig;
+
 /// Page cache behavior hint for file I/O (maps to posix_fadvise on Linux/Unix)
 pub use crate::data_loader::options::PageCacheMode;
+
+/// Unified storage configuration for all backends
+/// 
+/// Use this enum to configure backend-specific options when creating stores
+/// with `store_for_uri_with_config()`. This allows you to pass the correct
+/// configuration type for each storage backend (file://, direct://, etc.)
+#[derive(Debug, Clone)]
+pub enum StorageConfig {
+    /// Configuration for regular file:// URIs (page cache, range engine)
+    File(crate::file_store::FileSystemConfig),
+    /// Configuration for direct:// URIs and O_DIRECT operations
+    Direct(crate::file_store_direct::FileSystemConfig),
+}
 
 /// URI scheme detection
 pub use crate::object_store::{Scheme, infer_scheme};

--- a/src/file_store.rs
+++ b/src/file_store.rs
@@ -397,6 +397,7 @@ impl FileSystemObjectStore {
         
         // Apply page cache hint (use configured mode or Auto by default)
         let cache_mode = self.config.page_cache_mode.unwrap_or(PageCacheMode::Auto);
+        debug!("Applying page cache hint {:?} for file {} (size: {} bytes)", cache_mode, uri, file_size);
         let _ = apply_page_cache_hint(&std_file, cache_mode, file_size);
         
         // Convert back to tokio file and read
@@ -471,6 +472,8 @@ impl ObjectStore for FileSystemObjectStore {
         
         // Apply page cache hint (use configured mode, or Random for range reads by default)
         let cache_mode = self.config.page_cache_mode.unwrap_or(PageCacheMode::Random);
+        debug!("Applying page cache hint {:?} for range read {} offset={} length={:?} (file size: {} bytes)", 
+               cache_mode, uri, offset, length, file_size);
         let _ = apply_page_cache_hint(&std_file, cache_mode, file_size);
         
         let mut file = fs::File::from_std(std_file);

--- a/src/object_store.rs
+++ b/src/object_store.rs
@@ -45,7 +45,8 @@ use crate::s3_logger::global_logger;
 use crate::file_store::FileSystemObjectStore;
 
 // Expose enhanced FS adapter with O_DIRECT support
-use crate::file_store_direct::{ConfigurableFileSystemObjectStore, FileSystemConfig};
+use crate::file_store_direct::ConfigurableFileSystemObjectStore;
+
 
 // --- Azure ---------------------------------------------------
 use bytes::Bytes;
@@ -2439,38 +2440,42 @@ pub fn store_for_uri_with_logger(uri: &str, logger: Option<crate::s3_logger::Log
 }
 
 /// Enhanced factory that supports configuration options for file I/O
-pub fn store_for_uri_with_config(uri: &str, file_config: Option<FileSystemConfig>) -> Result<Box<dyn ObjectStore>> {
-    store_for_uri_with_config_and_logger(uri, file_config, None)
+pub fn store_for_uri_with_config(uri: &str, config: Option<crate::api::StorageConfig>) -> Result<Box<dyn ObjectStore>> {
+    store_for_uri_with_config_and_logger(uri, config, None)
 }
 
 /// Enhanced factory with configuration and optional logger support
 pub fn store_for_uri_with_config_and_logger(
     uri: &str, 
-    file_config: Option<FileSystemConfig>,
+    config: Option<crate::api::StorageConfig>,
     logger: Option<crate::s3_logger::Logger>
 ) -> Result<Box<dyn ObjectStore>> {
     use crate::object_store_logger::LoggedObjectStore;
     use std::sync::Arc;
     
-    let store: Box<dyn ObjectStore> = match infer_scheme(uri) {
-        Scheme::File => {
-            if let Some(config) = file_config {
-                ConfigurableFileSystemObjectStore::boxed(config)
-            } else {
-                FileSystemObjectStore::boxed()
-            }
+    let store: Box<dyn ObjectStore> = match (infer_scheme(uri), config) {
+        (Scheme::File, Some(crate::api::StorageConfig::File(cfg))) => {
+            Box::new(FileSystemObjectStore::with_config(cfg))
         }
-        Scheme::Direct => {
-            if let Some(config) = file_config {
-                ConfigurableFileSystemObjectStore::boxed(config)
-            } else {
-                ConfigurableFileSystemObjectStore::boxed_direct_io()
-            }
+        (Scheme::File, None) => {
+            FileSystemObjectStore::boxed()
         }
-        Scheme::S3 => S3ObjectStore::boxed(),
-        Scheme::Azure => AzureObjectStore::boxed(),
-        Scheme::Gcs => GcsObjectStore::boxed(),
-        Scheme::Unknown => bail!("Unable to infer backend from URI: {uri}"),
+        (Scheme::Direct, Some(crate::api::StorageConfig::Direct(cfg))) => {
+            ConfigurableFileSystemObjectStore::boxed(cfg)
+        }
+        (Scheme::Direct, None) => {
+            ConfigurableFileSystemObjectStore::boxed_direct_io()
+        }
+        (Scheme::Direct, Some(crate::api::StorageConfig::File(_))) => {
+            bail!("Cannot use FileSystemConfig with direct:// URI - use DirectFileSystemConfig instead")
+        }
+        (Scheme::File, Some(crate::api::StorageConfig::Direct(_))) => {
+            bail!("Cannot use DirectFileSystemConfig with file:// URI - use FileSystemConfig instead")
+        }
+        (Scheme::S3, _) => S3ObjectStore::boxed(),
+        (Scheme::Azure, _) => AzureObjectStore::boxed(),
+        (Scheme::Gcs, _) => GcsObjectStore::boxed(),
+        (Scheme::Unknown, _) => bail!("Unable to infer backend from URI: {uri}"),
     };
     
     // Wrap with logger if provided

--- a/tests/test_api_storage_config.rs
+++ b/tests/test_api_storage_config.rs
@@ -1,0 +1,206 @@
+// tests/test_api_storage_config.rs
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// SPDX-FileCopyrightText: 2025 Russ Fellows <russ.fellows@gmail.com>
+
+//! Tests for unified StorageConfig API (Issue #85)
+//!
+//! This test validates that the FileSystemConfig type mismatch is fixed and
+//! that users can configure page cache mode via the public API.
+
+use anyhow::Result;
+use s3dlio::api::{
+    FileSystemConfig, DirectFileSystemConfig, StorageConfig,
+    PageCacheMode, store_for_uri_with_config,
+};
+use s3dlio::range_engine_generic::RangeEngineConfig;
+use std::time::Duration;
+use tempfile::TempDir;
+use std::fs::File;
+use std::io::Write;
+
+#[tokio::test]
+async fn test_file_config_public_api() -> Result<()> {
+    // Create temp directory with test file
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Hello, world!")?;
+    drop(file);
+    
+    // Create FileSystemConfig with page cache mode
+    let config = FileSystemConfig {
+        enable_range_engine: false,
+        range_engine: Default::default(),
+        page_cache_mode: Some(PageCacheMode::Sequential),
+    };
+    
+    let uri = format!("file://{}", test_file.display());
+    
+    // This should compile and work now!
+    let store = store_for_uri_with_config(&uri, Some(StorageConfig::File(config)))?;
+    
+    // Verify it works by reading the file
+    let data = store.get(&uri).await?;
+    assert_eq!(&data[..], b"Hello, world!");
+    
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_file_config_random_mode() -> Result<()> {
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Random access test data")?;
+    drop(file);
+    
+    let config = FileSystemConfig {
+        enable_range_engine: false,
+        range_engine: Default::default(),
+        page_cache_mode: Some(PageCacheMode::Random),
+    };
+    
+    let uri = format!("file://{}", test_file.display());
+    let store = store_for_uri_with_config(&uri, Some(StorageConfig::File(config)))?;
+    
+    let data = store.get(&uri).await?;
+    assert_eq!(&data[..], b"Random access test data");
+    
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_file_config_auto_mode() -> Result<()> {
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Auto mode test")?;
+    drop(file);
+    
+    let config = FileSystemConfig {
+        enable_range_engine: false,
+        range_engine: Default::default(),
+        page_cache_mode: None,  // Auto mode (default)
+    };
+    
+    let uri = format!("file://{}", test_file.display());
+    let store = store_for_uri_with_config(&uri, Some(StorageConfig::File(config)))?;
+    
+    let data = store.get(&uri).await?;
+    assert_eq!(&data[..], b"Auto mode test");
+    
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_direct_config_not_for_file_uri() -> Result<()> {
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Test data")?;
+    drop(file);
+    
+    // Create DirectFileSystemConfig
+    let config = DirectFileSystemConfig {
+        direct_io: true,
+        alignment: 4096,
+        min_io_size: 4096,
+        sync_writes: false,
+        enable_range_engine: false,
+        range_engine: Default::default(),
+        buffer_pool: None,
+    };
+    
+    let uri = format!("file://{}", test_file.display());
+    
+    // This should fail with clear error message
+    let result = store_for_uri_with_config(&uri, Some(StorageConfig::Direct(config)));
+    
+    assert!(result.is_err());
+    if let Err(e) = result {
+        let err_msg = e.to_string();
+        assert!(err_msg.contains("Cannot use DirectFileSystemConfig with file://"));
+    }
+    
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_file_config_not_for_direct_uri() -> Result<()> {
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Test data")?;
+    drop(file);
+    
+    // Create FileSystemConfig
+    let config = FileSystemConfig {
+        enable_range_engine: false,
+        range_engine: Default::default(),
+        page_cache_mode: Some(PageCacheMode::Sequential),
+    };
+    
+    let uri = format!("direct://{}", test_file.display());
+    
+    // This should fail with clear error message
+    let result = store_for_uri_with_config(&uri, Some(StorageConfig::File(config)));
+    
+    assert!(result.is_err());
+    if let Err(e) = result {
+        let err_msg = e.to_string();
+        assert!(err_msg.contains("Cannot use FileSystemConfig with direct://"));
+    }
+    
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_no_config_still_works() -> Result<()> {
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("test.txt");
+    let mut file = File::create(&test_file)?;
+    file.write_all(b"Default config test")?;
+    drop(file);
+    
+    let uri = format!("file://{}", test_file.display());
+    
+    // Should work with None config
+    let store = store_for_uri_with_config(&uri, None)?;
+    let data = store.get(&uri).await?;
+    assert_eq!(&data[..], b"Default config test");
+    
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_range_engine_with_page_cache() -> Result<()> {
+    let temp_dir = TempDir::new()?;
+    let test_file = temp_dir.path().join("large.txt");
+    let mut file = File::create(&test_file)?;
+    
+    // Create a larger file (1MB) to potentially trigger range engine
+    let data = vec![b'A'; 1024 * 1024];
+    file.write_all(&data)?;
+    drop(file);
+    
+    let config = FileSystemConfig {
+        enable_range_engine: true,
+        range_engine: RangeEngineConfig {
+            chunk_size: 256 * 1024,           // 256KB chunks
+            max_concurrent_ranges: 4,
+            min_split_size: 512 * 1024,       // 512KB threshold
+            range_timeout: Duration::from_secs(60),
+        },
+        page_cache_mode: Some(PageCacheMode::Sequential),
+    };
+    
+    let uri = format!("file://{}", test_file.display());
+    let store = store_for_uri_with_config(&uri, Some(StorageConfig::File(config)))?;
+    
+    let result = store.get(&uri).await?;
+    assert_eq!(result.len(), 1024 * 1024);
+    assert_eq!(result[0], b'A');
+    
+    Ok(())
+}

